Framework/Algorithms/src/DiscusMultipleScatteringCorrection.cpp

```diff
@@ -219,46 +369,53 @@ void DiscusMultipleScatteringCorrection::exec() {
-    MatrixWorkspace_sptr SQWSPoints = API::WorkspaceFactory::Instance().create(SQWS, SQWS->getNumberHistograms(),
-                                                                               SQWS->blocksize(), SQWS->blocksize());
-    SQWSPoints->setSharedY(0, SQWS->sharedY(0));
-    SQWSPoints->setSharedE(0, SQWS->sharedE(0));
-    std::vector<double> newX = SQWS->histogram(0).dataX();
-    newX.pop_back();
-    SQWSPoints->setSharedX(0, HistogramData::Points(newX).cowData());
-    SQWS = SQWSPoints;
-  }
+
+  m_SQWS = getProperty("StructureFactorWorkspace");
+  prepareStructureFactor();
 
   MatrixWorkspace_sptr sigmaSSWS = getProperty("ScatteringCrossSection");
   if (sigmaSSWS)
-    m_sigmaSS = std::make_shared<HistogramData::Histogram>(sigmaSSWS->histogram(0));
-
-  double lambdamin, lambdamax;
-  getXMinMax(*inputWS, lambdamin, lambdamax);
-  m_SQHist = std::make_shared<HistogramData::Histogram>(SQWS->histogram(0));
-  auto QSQHist = prepareQSQ(2 * M_PI / lambdamin);
-
-  const auto inputNbins = static_cast<int>(inputWS->blocksize());
-  int nlambda = getProperty("NumberOfWavelengthPoints");
-  if (isEmpty(nlambda) || nlambda > inputNbins) {
-    if (!isEmpty(nlambda)) {
-      g_log.warning() << "The requested number of wavelength points is larger "
-                         "than the spectra size. "
-                         "Defaulting to spectra size.\n";
+    m_sigmaSS = std::make_shared<DataObjects::Histogram1D>(sigmaSSWS->getSpectrum(0));
+
+  // for inelastic we could calculate the qmax based on the min\max w in the S(Q,w) but that
+  // would bake as assumption that S(Q,w)=0 beyond the limits of the supplied data
+  double qmax = std::numeric_limits<float>::max();
+  EFixedProvider efixed(*inputWS);
+  m_EMode = efixed.emode();
+  g_log.information("EMode=" + DeltaEMode::asString(m_EMode) + " detected");
+  if (m_EMode == Kernel::DeltaEMode::Elastic) {
+    double kmin, kmax;
+    getXMinMax(*inputWS, kmin, kmax);
+    qmax = 2 * kmax;
+  }
+  m_QSQWS = prepareQSQ(qmax);
+  size_t totalPoints = 0;
+  for (size_t i = 0; i < m_QSQWS->getNumberHistograms(); i++) {
+    totalPoints += m_QSQWS->histogram(i).size();
+  }
+
+  m_simulateEnergiesIndependently = getProperty("SimulateEnergiesIndependently");
+  // call this function with dummy efixed to determine total possible simulation points
+  const auto inputNbins = generateInputKOutputWList(-1.0, inputWS->points(0).rawData()).size();
+
+  int nSimulationPointsInt = getProperty("NumberOfSimulationPoints");
+  size_t nSimulationPoints = static_cast<size_t>(nSimulationPointsInt);
+  if (isEmpty(nSimulationPoints) || nSimulationPoints > inputNbins) {
+    if (!isEmpty(nSimulationPoints)) {
+      g_log.warning() << "The requested number of simulation points is larger "
+                         "than the maximum number of simulations per spectra. "
+                         "Defaulting to "
+                      << inputNbins << ".\n ";
```

Comment:
If the simulation points are empty we don't emit a warning?
If we don't warn for an empty number of points we can simplify the if.

If we should check a better flow would be:

```c++
if (isEmpty(nSimulationPoints)){
    g_log.warning("Empty")
    nSimulationPoints = inputNbins;
} else if (nSimulationPoints > inputNbins) {
    g_log.warning(" The requested number...")
    nSimulationPoints = inputNbins;  // Intentionally duplicated as simpler flow > DRY principle for 2 statements
}
```

Reply:
nSimulationPoints is an optional parameter so fine not to supply it and no warning required in that case

Reply:
Can we simplify the if from:
```c++
  if (isEmpty(nSimulationPoints) || nSimulationPoints > inputNbins) {
    if (!isEmpty(nSimulationPoints)) {
```
To:
```c++
if (isEmpty(nSimulationPoints)){ 
    nSimulationPoints = inputNbins; 
} else if (nSimulationPoints > inputNbins) { 
    g_log.warning(" The requested number...") 
    nSimulationPoints = inputNbins;
}
```